v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/utils/ShadowTableCreator.java

```diff
@@ -0,0 +1,182 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.templates.utils;
+
+import com.google.api.gax.longrunning.OperationFuture;
+import com.google.cloud.spanner.BatchClient;
+import com.google.cloud.spanner.BatchReadOnlyTransaction;
+import com.google.cloud.spanner.DatabaseAdminClient;
+import com.google.cloud.spanner.Dialect;
+import com.google.cloud.spanner.TimestampBound;
+import com.google.cloud.teleport.v2.spanner.ddl.Column;
+import com.google.cloud.teleport.v2.spanner.ddl.Ddl;
+import com.google.cloud.teleport.v2.spanner.ddl.IndexColumn;
+import com.google.cloud.teleport.v2.spanner.ddl.InformationSchemaScanner;
+import com.google.cloud.teleport.v2.spanner.ddl.Table;
+import com.google.cloud.teleport.v2.spanner.type.Type;
+import com.google.spanner.admin.database.v1.UpdateDatabaseDdlMetadata;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerAccessor;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerConfig;
+
+/** Helper class to create shadow tables for different source types. */
+public class ShadowTableCreator {
+
+  private final SpannerAccessor spannerAccessor;
+  private final SpannerAccessor metadataSpannerAccessor;
+  private final Dialect dialect;
+  private final SpannerConfig spannerConfig;
+  private final SpannerConfig metadataConfig;
+  private String shadowTablePrefix;
+
+  public ShadowTableCreator(
+      SpannerConfig spannerConfig,
+      SpannerConfig metadataConfig,
+      Dialect dialect,
+      String shadowTablePrefix) {
+    this.spannerAccessor = SpannerAccessor.getOrCreate(spannerConfig);
+    ;
+    this.metadataSpannerAccessor = SpannerAccessor.getOrCreate(metadataConfig);
+    this.dialect = dialect;
+    this.spannerConfig = spannerConfig;
+    this.metadataConfig = metadataConfig;
+    this.shadowTablePrefix = shadowTablePrefix;
+  }
+
+  public void createShadowTablesInSpanner() {
+
+    BatchClient batchClient = spannerAccessor.getBatchClient();
+    BatchReadOnlyTransaction context =
+        batchClient.batchReadOnlyTransaction(TimestampBound.strong());
+    InformationSchemaScanner scanner = new InformationSchemaScanner(context, dialect);
+    Ddl informationSchema = scanner.scan();
+    List<String> dataTablesWithoutShadowTables = getDataTablesWithNoShadowTables(informationSchema);
```

Comment:
Maybe we can use CREATE TABLE IF NOT EXISTS to simplify this logic? It also avoids race condition when multiple workers trying to create shadow tables.

Reply:
Shadow tables are created in the main class before the workers are spawned, so there will not be race condition. I could not find programmatic way to doing create table if not exists, saw in the DatastreamToSpanner template - the handling is similar.